template/src/main/java/com/asyncapi/service/MessageHandlerService.java

```diff
@@ -1,19 +1,50 @@
 package com.asyncapi.service;
+{%- set hasSubscribe = false -%}
+{%- set hasPublish = false -%}
+{%- for channelName, channel in asyncapi.channels() -%}
+    {%- if channel.hasPublish() -%}
+        {%- set hasPublish = true -%}
+    {%- endif -%}
+    {%- if channel.hasSubscribe() -%}
+        {%- set hasSubscribe = true -%}
+    {%- endif -%}
+{%- endfor %}
 
 import org.springframework.messaging.Message;
 import org.springframework.stereotype.Service;
-
+{% if asyncapi | isProtocol('kafka') and hasPublish %}
+import org.springframework.kafka.annotation.KafkaListener;
+import org.springframework.kafka.support.KafkaHeaders;
+import org.springframework.messaging.handler.annotation.Header;
+import org.springframework.messaging.handler.annotation.Payload;
+    {% for channelName, channel in asyncapi.channels() %}
+        {%- if channel.hasPublish() %}
+import com.asyncapi.model.{{channel.publish().message().payload().uid() | camelCase | upperFirst}};
+        {% endif -%}
+    {% endfor -%}
+{% endif %}
 @Service
 public class MessageHandlerService {
-
-{% for topicName, topic in asyncapi.topics %}
-  {% if topic.subscribe %}
-    public void handle{{topic.x-service-name | upperFirst}}(Message<?> message) {
-        System.out.println("handler {{topic.x-service-name}}");
+{% if asyncapi | isProtocol('kafka') %}
+    {% for channelName, channel in asyncapi.channels() %}
+        {%- if channel.hasPublish() %}
+    @KafkaListener(topics = "{{channelName}}"{% if channel.publish().binding('kafka') %}, groupId = "{{channel.publish().binding('kafka').groupId}}"{% endif %})
+    public void {{channel.publish().id() | camelCase}}(@Payload {{channel.publish().message().payload().uid() | camelCase | upperFirst}} payload,
+                       @Header(KafkaHeaders.RECEIVED_MESSAGE_KEY) Integer key,
```

Comment:
It would be interesting if we could extract also Kafka Headers here

Reply:
there is example with Key header, this is template generation, not the documentation of spring-kafka library.

Reply:
Fair enough, we can add them later.